shacl12-rules/index.html

```diff
@@ -867,942 +951,1014 @@ <h2>Shapes Rules Language Grammar</h2>
               </tr>
 
               <tr style="vertical-align: baseline">
-                <td><code>[45]&nbsp;&nbsp;</code></td>
+                <td><code>[46]&nbsp;&nbsp;</code></td>
                 <td><code><span class="doc-ref" id="rObjectPath">ObjectPath</span></code></td>
                 <td>&nbsp;&nbsp;::=&nbsp;&nbsp;</td>
                 <td><code class="gRuleBody"><a href="#rGraphNodePath">GraphNodePath</a> <a href="#rAnnotationPath">AnnotationPath</a></code></td>
               </tr>
 
               <tr style="vertical-align: baseline">
-                <td><code>[46]&nbsp;&nbsp;</code></td>
+                <td><code>[47]&nbsp;&nbsp;</code></td>
                 <td><code><span class="doc-ref" id="rPath">Path</span></code></td>
                 <td>&nbsp;&nbsp;::=&nbsp;&nbsp;</td>
                 <td><code class="gRuleBody"><a href="#rPathSequence">PathSequence</a></code></td>
               </tr>
 
               <tr style="vertical-align: baseline">
-                <td><code>[47]&nbsp;&nbsp;</code></td>
+                <td><code>[48]&nbsp;&nbsp;</code></td>
                 <td><code><span class="doc-ref" id="rPathSequence">PathSequence</span></code></td>
                 <td>&nbsp;&nbsp;::=&nbsp;&nbsp;</td>
                 <td><code class="gRuleBody"><a href="#rPathEltOrInverse">PathEltOrInverse</a> ( <span class="token">'/'</span> <a href="#rPathEltOrInverse">PathEltOrInverse</a> )*</code></td>
               </tr>
 
               <tr style="vertical-align: baseline">
-                <td><code>[48]&nbsp;&nbsp;</code></td>
+                <td><code>[49]&nbsp;&nbsp;</code></td>
                 <td><code><span class="doc-ref" id="rPathEltOrInverse">PathEltOrInverse</span></code></td>
                 <td>&nbsp;&nbsp;::=&nbsp;&nbsp;</td>
                 <td><code class="gRuleBody"><a href="#rPathElt">PathElt</a> | <span class="token">'^'</span> <a href="#rPathElt">PathElt</a></code></td>
               </tr>
 
               <tr style="vertical-align: baseline">
-                <td><code>[49]&nbsp;&nbsp;</code></td>
+                <td><code>[50]&nbsp;&nbsp;</code></td>
                 <td><code><span class="doc-ref" id="rPathElt">PathElt</span></code></td>
                 <td>&nbsp;&nbsp;::=&nbsp;&nbsp;</td>
                 <td><code class="gRuleBody"><a href="#rPathPrimary">PathPrimary</a></code></td>
               </tr>
 
               <tr style="vertical-align: baseline">
-                <td><code>[50]&nbsp;&nbsp;</code></td>
+                <td><code>[51]&nbsp;&nbsp;</code></td>
                 <td><code><span class="doc-ref" id="rPathPrimary">PathPrimary</span></code></td>
                 <td>&nbsp;&nbsp;::=&nbsp;&nbsp;</td>
                 <td><code class="gRuleBody"><a href="#riri">iri</a> | <span class="token">'a'</span> | <span class="token">'('</span> <a href="#rPath">Path</a> <span class="token">')'</span> </code></td>
               </tr>
```

Comment:
this does only cover sequence `/` and inverse `^` directly.  but not others like `|, +, *, or !`. 

Although they could be expressed/dealt with differently:

Missing Construct | Logic Type | Replacement Mechanism in SHACL Rules
-- | -- | --
Alternative (\|) | Disjunction | Multiple Rules
One-or-more (+) | Recursion | Recursive Rules (Base case + Step)
Zero-or-more (*) | Recursion + Identity | Recursive Rules + BIND (Identity)
Negated Set (!) | Negation | Wildcard Predicate (?p) + FILTER

So was this restriction in the grammar deliberate to move complexity out of the property path string and **into the rule structure** itself?

Reply:
It's deliberate. Sequence and inverse are syntactic sugar for a triple pattern.

The others introduce path operators. I could not see that they will be analysable for rule dependency calculations, and even if they are, whether they need well-formness constraints.

Negated set in particular looks in conflict with any datalog negation because the set of predicated is not bounded.

As you note, there are rules that have similar effects.

So - an area to consider later.

Note also: `TRANSITIVE` - a short form of writing transitive properties. Clarity for the rule users; easier for an engine to provide a custom implementation. If simple engines may not implement full rules, it gives one case of positive recursion without needing full support. Also `SYMMETRIC` and `INVERSE`.